lib/wallet/src/Cardano/Wallet/Write/Tx/Sign.hs

```diff
@@ -188,12 +188,22 @@ estimateKeyWitnessCount utxo txbody@(Cardano.TxBody txbodycontent) =
             fromIntegral
             $ sumVia estimateMaxWitnessRequiredPerInput
             $ mapMaybe toTimelockScript scripts
+        -- when wallets uses reference input it means script containing
+        -- its policy key was already published in previous tx
+        -- if so we need to add one witness that will stem from policy signing
+        -- key. As it is not allowed to publish and consume in the same transaction
+        -- we are not going to double count.
+        txRefInpsWit =
+            case Cardano.txInsReference txbodycontent of
+                Cardano.TxInsReferenceNone -> 0
+                Cardano.TxInsReference _ _ -> 1
```

Comment:
This seems like it's assuming each reference input points to an inline script which requires 1 key witness?

Reply:
if there is reference input(s) then each reference input refers to tx where script was published. Each script has structure : `[cosigner#0, timelocks]` and it means it contains policy key. We use ONE policy key per wallet, which means there is going to be ONE witness from this or ZERO. So Whatever number of reference inputs,one witness with wallet's policy signing key is going to be used. 

Remember that we are here in SHELLEY and script regulates minting/burning. So yes, we will only have ONE cosigner. In shared style it would be different but atm we are not supporting minting/burning for shared wallets.

Reply:
In the presence of Plutus validator scripts, the number of extra witnesses can arbitrarily depend on the data in the reference inputs, so the estimate can always fail terribly in principle.

But even in the absence of Plutus validator scripts, I think that the presence of reference inputs alone is not a good basis for estimation. 🤔 What about combining this with `Cardano.txMintValue`? The logic being that if the transaction mints something, and there is a reference input, then the most likely of that input is to contain a script which requires one additional verification key.

```
      case Cardano.txInsReference txbodycontent of
          Cardano.TxInsReferenceNone -> 0
          Cardano.TxInsReference{} ->
              case CardanotxMintValue txbodycontent of
                  Cardano.TxMintValueNone -> 0
                  Cardano.TxMintValue{} -> 1
```

Reply:
thanks! Very good suggestion! Adopted!